### ISSUE TITLE
[Feature request] Support package snapshot

### ISSUE THREAD

**Issue by weihanglo** [wbthomason/packer.nvim#298]
Sometimes upgrading packages may results in conflicts between packages. By having a snapshot of all current packages with exact commits, we can restore from the snapshot without much labor.

Reference: [vim-plug](https://github.com/junegunn/vim-plug/blob/cffcfe150bda75177920530c5f1911b52f70a065/doc/plug.txt#L201) supports snapshot.

**Comment by wbthomason** [comment]
Thanks for the idea and reference! I'd be happy to help guide a PR implementing this (it should be pretty straightforward), if you or anyone else is interested.

**Comment by weihanglo** [comment]
Yes, I can help! It would be awesome to have some instruction.

**Comment by c3n21** [comment]
Since already a month has passed and there isn't any progress I would like to step-in and make this feature (since I need this too).

The idea I had is the following:
- create a **PackerSnapshot** command which can take an optional argument which will be the name of the snapshot (all the plugins will be snapshotted except the disabled ones)
- create a **PackerRollback** command which takes an argument and what it does is rollback to the given snapshot (with autocompletion for the snapshots' names which is nice)
- create a **PackerDelete** command which deletes a snapshot given the name
- in the config we could add an option to enable automatic snapshot after each **PackerInstall** or **PackerUpdate** or **PackerClean**
- eventually add a timer which deletes old snapshots created automatically after installs, updates and cleans (so snapshots created with a name will be saved permanently unless the user wants to delete it)

How does this sound to you @wbthomason @weihanglo ?

**Comment by bew** [comment]
_me, dropping in from nowhere_

Imo this gives too much intelligence to the snapshot system.

Taking inspiration from vim-plug, I like how it only generates a simple vim file with 1 line for each plugin, specifying it's name and the revision.
And the file content is in a scratch buffer, it's not saved anywhere the plugin knows about. If I want to keep it I can hit `:w some_place.vim` and manage the snapshot file however I want.

(here is an example snapshot: https://github.com/bew/dotfiles/blob/0d74848b7d8fde734c2ad47e7b4e78e27f43a3b0/nvim/plugins.lock) 

And when you want to restore a snapshot, you simply have to `:source some_place.vim` and run PlugInstall, this will source the snapshot and fill a special vim dictionary, that is used first when installing packages and choosing a revision.

What do you think of this?

**Comment by c3n21** [comment]
> _me, dropping in from nowhere, hi! wave_
> 
> Imo this gives too much intelligence to the snapshot system.
> 
> Taking inspiration from vim-plug, I like how it only generates a simple vim file with 1 line for each plugin, specifying it's name and the revision.
> And the file content is in a scratch buffer, it's not saved anywhere the plugin knows about. If I want to keep it I can hit `:w some_place.vim` and manage the snapshot file however I want.
> 
> (here is an example snapshot: https://github.com/bew/dotfiles/blob/0d74848b7d8fde734c2ad47e7b4e78e27f43a3b0/nvim/plugins.lock)
> 
> And when you want to restore a snapshot, you simply have to `:source some_place.vim`, this will source the snapshot and fill a special vim dictionary, and then force update all the plugins to the versions in the snapshot.
> 
> **In packer's case** the snapshot file could be in Lua, but the general idea could be kept I think. Unless there are specific things for packer? (_I don't actually use it currently, but I'm keeping my third eye on it I'd say_)
> 
> What do you think of this?

I was actually happy with that system since I've been using VimPlug for a while and I loved that snapshot feature, for me it would be fine.

However I feel like packer is not as **minimalist** as vim-plug since it can manage config files, extensions' dependencies and other features that are not strictly necessary for a plugin manager.
I believe that packer resembles more something like a package manager, and having a builtin API to manage snapshots would make sense for me (or maybe support them via extension?)

**Comment by weihanglo** [comment]
I have an almost identical opinion as what @bew's said. A dumb snapshot system is great enough to me. It is also simpler to implement. Therefore I propose that at this moment the snapshot system aligns to a simple design, and we'll then wait for feedback to decide the next iteration.

**Comment by bew** [comment]
@wbthomason Hi, what's your take on this? dumb/simple snapshot system vs a 'complex' one?

_Personally the dumb system is all I want, and the complex one would make me fear it in a way._

**Comment by wbthomason** [comment]
@bew There's a bit more discussion on #370, but: I favor a sort of middle ground, at least to start. I think we want snapshots to be closer to `plug`'s "list of commits" (though this may change later if we want to persist configuration variables, etc.), and I'm fine with having a set of `PackerSnapshot`, `PackerRollback`, `PackerSnapshotDelete` commands to make management a bit easier, but I don't think we want to add more complexity like automatic timed deletion, etc. - really, nothing more complex than handling saving the snapshot file for you.

All this said, I never used the snapshot feature (in `plug` or anything else) when I used other plugin managers, so I may not have a good intuition for what complexity is necessary/useful and what just adds bloat.

**tl;dr**: I think we want `plug`-like simple snapshots, but I'm fine also adding management commands (without any automatic management like auto-deletion).
